docs/osf-trolling-list/build_process.md

```diff
@@ -0,0 +1,127 @@
+  able to confirm the same binary can be generated in future. If we know given
+  toolchain reproduce binary bit by bit we are sure it also reproduce
+  software/firmware behavior. In that light making small incremental
+  improvement using toolchain that give reproducible results give us higher
+  chance that we will not introduce uncontrolled change in behavior and code
+  change will be reflected by compilation process without affecting previous
+  software behavior.
+
+# [OSFI0011] Docker containers as build systems "considered harmful"
+
+Docker containers for couple last years become default method for provide
+reproducible runtime environment for software as well those started to be very
+useful for developers to transfer exact configuration for building given
+software stack. We wrote about usefulness of Docker containers in embedded
+software development environment on 
+[3mdeb blog](https://blog.3mdeb.com/2018/2018-09-27-optimize-performance-in-docker/).
+
+After long time of using Docker containers for embedded software development
+and build environments we noticed problems for long time maintenance and
+reproducibility.
+
+## coreboot-sdk problems
+
+Example can be `coreboot-sdk` used for building [coreboot project](https://coreboot.org).
+
+Good summary of the problem was provided by [Thrilleratplay in guix-docker repo](https://github.com/Thrilleratplay/guix-docker#the-problem):
+
+> Currently, the coreboot build environment, `coreboot-sdk`, uses a Debian docker
+> base image.  To install additional required packages, `apt-get update` must be
+> run.  The resulting docker image is hosted in docker hub repository to be
+> retrieved at any time in the future.  However, at any time in the future,
+> building the same docker file will generate a different image based on the
+> latest packages used in apt-get.  Overtime, as packages are updated due to bug,
+> security or feature improvements, the provenance of the docker image in the
+> docker hub repository becomes increasingly difficult, if not impossible to
+> audit and reproduce.
+
+Some more elaborate notes related to `coreboot-sdk`
+
+1. There is no meaningful versioning of SDK
+   [[1]](https://hub.docker.com/r/coreboot/coreboot-sdk/tags) - at some point
+   there was versioning 1.32-53, now switched to git SHA. Both seem to have no
+   meaning and were released at arbitrary points in time. Result is that it is
+   hard to find which version of SDK works with version of coreboot tree.
+2. Validation of coreboot-sdk is not sufficient - for example recently python
+   fixes were merged [[2]](https://review.coreboot.org/c/coreboot/+/45265), somehow this change passed all QA checks, but SeaBIOS
+   use python (not python2 or python3) and all builds using this version of SDK
+   that compile SeaBIOS fail.
+3. coreboot-sdk is based on moving target Debian sid, it is close to impossible
+   to build the same Docker images at 2 different points in time.
```

Comment:
Could build a custom Docker image with all the required packages preinstalled, and then distribute it. Or just a Virtual Machine instead of a Docker.

Reply:
Virtual Machine is to some extent solution but not very scalable, eg. what will all CI/CD systems? Using VM is slow, but last resort it could be a solution that can work for a longer time. I'm more worried now if in long run we can reproducibly build binaries? If for any particular version of firmware, to achieve reproducibility, we have to have VM that would be madness.

`coreboot-sdk` is distributed on Docker Hub and can be downloaded, but it is not reproducible.

Reply:
@pietrushnic I expect that a need to release a new VM because i.e. a coreboot's build toolchain became incompatible with some older gcc - won't happen too often - so a set of VMs with preinstalled packages won't have to be too big, and the older VMs could be archived since they'd be rarely needed. For example, in case a coreboot 4.2 couldn't be built in a fresh VM, how often do you need a coreboot 4.2?

Reply:
@mikebdp2 this is not so obvious we still building 4.0.x: https://pcengines.github.io/
I know the argument that old stuff should be not used, but the reality is that customers can always come back with old stuff and we have to deal with that without breaking firmware e.g. we had a case when coreboot was in medical robot or in train.